File: shared/typescript/src/consts.ts
// consts/defaults + consts/errors: request/file size caps, sniff length, the
// canonical error strings the contract asserts, and the shared error-shaping
// helper. These have no dependencies on the rest of the package.

// ── consts/defaults ─────────────────────────────────────────────────────────
export const MAX_REQUEST_BYTES: number = 10 * 1024 * 1024; // 10 MB
export const MAX_FILE_BYTES: number = 1 << 20; // 1MB
export const SNIFF_LEN = 512;
export const NULL_BYTE = 0x00;
export const DEFAULT_LIMIT = 10;

// ── consts/errors ───────────────────────────────────────────────────────────
export const INVALID_JSON_BODY = "invalid JSON body";
export const INVALID_FORM_DATA = "invalid form data";
export const EXPECTED_FORM_CONTENT_TYPE =
  "expected content-type: application/x-www-form-urlencoded or multipart/form-data";
export const INVALID_MULTIPART = "invalid multipart form data";
export const EXPECTED_MULTIPART_CONTENT_TYPE = "expected content-type: multipart/form-data";
export const FILE_NOT_FOUND = "file not found in form data";
export const NOT_FOUND = "not found";
export const FILE_SIZE_EXCEEDS = "file size exceeds limit";
export const ONLY_TEXT_PLAIN = "only text/plain files are allowed";
export const FILE_NOT_TEXT = "file does not look like plain text";
export const INTERNAL_ERROR = "internal error";
// Web-suite error strings (contract/web.json). INVALID_TOKEN covers every
// /jwt/verify rejection (missing/malformed/bad-signature/expired); VALIDATION_FAILED
// is the /validate 400; INVALID_N is the /compute boundary rejection (house
// "invalid <thing>" style).
export const INVALID_TOKEN = "invalid token";
export const VALIDATION_FAILED = "validation failed";
export const INVALID_N = "invalid n";

export type ErrorResponse = { error: string; details?: string };

export function makeError(error: string, detail?: unknown): ErrorResponse {
  if (detail instanceof Error) {
    return detail.message ? { error, details: detail.message } : { error };
  }
  if (typeof detail === "string" && detail) {
    return { error, details: detail };
  }
  return { error };
}

File: shared/rust/src/consts.rs
//! Cross-server constants: canonical error strings and request limits.
//!
//! These mirror `shared/go/consts` (and the TS `@bench/shared` consts) verbatim
//! so every framework emits byte-identical error bodies — the contract asserts
//! them exactly.

// Error `error` field strings (the `{"error": ...}` value).
pub const ERR_INVALID_JSON: &str = "invalid JSON body";
pub const ERR_REQUEST_TOO_LARGE: &str = "request body too large";
pub const ERR_INVALID_FORM: &str = "invalid form data";
pub const ERR_EXPECTED_FORM_CONTENT_TYPE: &str =
    "expected content-type: application/x-www-form-urlencoded or multipart/form-data";
pub const ERR_INVALID_MULTIPART: &str = "invalid multipart form data";
pub const ERR_EXPECTED_MULTIPART_CONTENT_TYPE: &str = "expected content-type: multipart/form-data";
pub const ERR_FILE_NOT_FOUND: &str = "file not found in form data";
pub const ERR_FILE_SIZE_EXCEEDED: &str = "file size exceeds limit";
pub const ERR_INVALID_FILE_TYPE: &str = "only text/plain files are allowed";
pub const ERR_NOT_PLAIN_TEXT: &str = "file does not look like plain text";
pub const ERR_NOT_FOUND: &str = "not found";
pub const ERR_INTERNAL: &str = "internal error";

// Web suite (PLAN §5) — house "invalid <thing>" / "<what> failed" style.
pub const ERR_INVALID_TOKEN: &str = "invalid token";
pub const ERR_VALIDATION_FAILED: &str = "validation failed";
pub const ERR_INVALID_N: &str = "invalid n";

/// Global request-body cap (10 MiB) applied to every route.
pub const MAX_REQUEST_BYTES: usize = 10 << 20;
/// Per-file upload cap (1 MiB) enforced on `POST /params/file`.
pub const MAX_FILE_BYTES: usize = 1 << 20;
/// Bytes inspected when sniffing an upload's content type.
pub const SNIFF_LEN: usize = 512;
/// Default `limit` when the query param is missing or unparseable.
pub const DEFAULT_LIMIT: i64 = 10;
